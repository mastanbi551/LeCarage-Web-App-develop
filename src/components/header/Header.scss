@import "../../styles/variables.scss";
.main-nav {
  padding: 0px;
  background: $primaryGreen !important;
  height: 70px;

  &__logo {
    width: 52px;
    height: 50px;
  }

  @media (max-width: 767px) {
    &__notofication-icon {
      position: absolute;
      right: 60px;
      color: $primaryWhite;
    }
    .container-fluid {
      position: relative;
    }
  }

  @media (min-width: 768px) {
    .navbar-expand-md {
      .navbar-collapse {
        display: block !important;
      }
      .navbar-nav {
        float: right;
      }
    }
    .container-fluid {
      padding-left: 90px;
    }
  }

  .navbar-nav li,
  .navbar-nav .dropdown {
    display: inline-block;
  }
  a.nav-link {
    color: $primaryWhite;
    padding-right: 10px;
  }
  .nav-arrow-icon {
    margin-left: 10px;
  }
  .navbar-nav .dropdown.show {
    .nav-arrow-icon {
      transform: rotate(180deg);
    }
  }
  .navbar-nav .nav-link.active,
  .navbar-nav .show > .nav-link {
    color: $primaryWhite;
  }
  .navbar-expand-md .navbar-nav .dropdown-menu {
    padding: 0px;
  }
  .user-icon {
    background: $darkgrey;
  }
  .dropdown-menu.show {
    width: 280px;
  }
}

.new {
  white-space: pre-wrap;
}

.fixed {
  position: fixed;
  top: 0px;
  width: 100%;
  z-index: 1000;
}
