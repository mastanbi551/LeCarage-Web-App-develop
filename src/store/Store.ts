import {configureStore} from '@reduxjs/toolkit';
import loginReducer from '../pages/login/loginState/LoginSlice';
import headerReducer from '../components/header/headerstate/HeaderSlice';
import sidebarReducer from '../components/sidebar/sidebarState/SidebarSlice';
import forgotPasswordReducer from '../pages/forgotPassword/forgotPasswordState/ForgotPasswordSlice';
import otpVerifyReducer from '../pages/otpVerification/otpState/OtpVerifySlice';
import resetPasswordReducer from '../pages/resetPassword/resetPasswordState/ResetPasswordSlice';
import commonReducer from '../pages/commonStates/CommonSlice';
import userManagementReducer from '../pages/userManagement/userManagementPage/userManagementState/UserManagementSlice';
import carManagementReducer from '../pages/userManagement/carManagementPage/carManagementState/CarManagementSlice';
import userTypeManagementReducer from '../pages/userManagement/userTypeManagementPage/userTypeManagementPageState/UserTypeManagementSlice';
import customerManagementReducer from '../pages/userManagement/customerManagementPage/customerManagementPageState/CustomerManagementSlice';
import bookingManagementReducer from '../pages/bookingManagement/bookingsManagementPageState/BookingsManagementSlice';
import createSagaMiddleware from 'redux-saga';
import Saga from '../middleWare/Saga';
const sagaMiddleware = createSagaMiddleware();
export const Store = configureStore({
  reducer: {
    loginReducer,
    headerReducer,
    sidebarReducer,
    forgotPasswordReducer,
    resetPasswordReducer,
    otpVerifyReducer,
    commonReducer,
    userManagementReducer,
    carManagementReducer,
    userTypeManagementReducer,
    customerManagementReducer,
    bookingManagementReducer
  },
  middleware: [sagaMiddleware],
});
sagaMiddleware.run(Saga);
export type AppDispatch = typeof Store.dispatch;
export type RootState = ReturnType<typeof Store.getState>;