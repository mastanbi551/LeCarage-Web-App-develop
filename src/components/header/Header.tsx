import React, { useEffect, useState } from 'react';
import './Header.scss';
import { LeCarageLogo } from '../leCarageLogo/LeCarageLogo';
import { ImagesPath } from '../../utils/constants';
import { BsBell } from 'react-icons/bs';
import { IoIosArrowDown } from 'react-icons/io';
import { FiLogOut } from 'react-icons/fi';
import { SlEqualizer } from 'react-icons/sl';
import { AiOutlineUser } from 'react-icons/ai';
import { ModalDropDown } from '../modalDropDown/ModalDropDown';
import {
  Navbar,
  NavbarBrand,
  Nav,
  NavItem,
  NavLink,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
} from 'reactstrap';
import { useNavigate } from 'react-router';
import { useAppDispatch, useAppSelector } from '../../hooks/Hooks';
import { onLoggedOut, reset } from '../../pages/login/loginState/LoginSlice';
import {
  userRoleType,
  adminUserProfile,
  defaultUserType,
  headerReset,
  defaultDetailsData,
} from './headerstate/HeaderSlice';
import { onApiCall } from '../../components/header/headerstate/HeaderSlice';
import { Details } from './headerstate/HeaderStateInterface';
import { responseStatus } from '../../utils/constants/ConstantStrings';
interface IProps {
  setAuth: (user: boolean) => void;
}

export const Header: React.FC<IProps> = ({ setAuth }) => {
  const dispatch = useAppDispatch();
  const isAuthenticate = localStorage.getItem('isAuthenticate');
  const userId = localStorage.getItem('userId');
  const parseJSON = isAuthenticate && JSON.parse(isAuthenticate);
  const uID = userId && JSON.parse(userId);
  const {
    userRoleList,
    responseData,
    isLoading,
    defaultDetails,
  } = useAppSelector((state) => state.headerReducer);
  const [showLocationRole, setShowLocationRole] = useState(false);

  const profileInfo = userRoleList;
  const apiparams = {
    uId: uID,
  };

  const [userProfile, setUserProfile] = useState('');
  const [warehouses, setWarehouses] = useState<string[]>([]);
 
  useEffect(() => {
    if (defaultDetails && defaultDetails.userTypeLabel !== '') {
      setUserProfile(
        `${defaultDetails.userTypeLabel} => ${defaultDetails.warehouseLabel}`
      );
    }
  }, [defaultDetails]);

  useEffect(() => {
    if (isLoading) {
      if (responseData.status === responseStatus.success) {
        if(responseData.details.length > 0){
          const DefaultResponseData = responseData.details.filter(
            (item: Details) => item.isDefault === true
          );
          dispatch(userRoleType(responseData.details[0].userTypeValue));
          dispatch(adminUserProfile(responseData.details));
          dispatch(defaultDetailsData(DefaultResponseData[0]));
          dispatch(
            defaultUserType(
              responseData.details.filter((item:Details) => item.isDefault === true)
            )
          );
          setWarehouses(
            responseData.details
              .map((item: Details) => item.warehouseLabel)
              .filter(
                (value: string, index: number, self: any) =>
                  self.indexOf(value) === index
              )
          );
        }
       
      }
    }
  }, [isLoading, responseData]);

  useEffect(() => {
    if (userRoleList.length === 0 ) {
      dispatch(onApiCall(apiparams));
    }
  }, [userRoleList]);

  const navigate = useNavigate();

  const handleUserProfile = (value: Details) => {
    setUserProfile(`${value.userTypeLabel} -> ${value.warehouseLabel}`);
    dispatch(defaultDetailsData(value));

    navigate('/');
  };

  const logOutOnClick = () => {
    if (!parseJSON.isRemember) {
      localStorage.removeItem('isAuthenticate');
    } else {
      const newJson = { ...parseJSON, isAuthenticate: false };
      localStorage.setItem('isAuthenticate', JSON.stringify(newJson));
    }
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
    setAuth(false);
    dispatch(headerReset());
    dispatch(reset());
    navigate('/');
  };

  return (
    <div className='fixed'>
      <Navbar className='main-nav' color='light' light expand='md'>
        <NavbarBrand href='/'>
          <LeCarageLogo
            source={ImagesPath.lecaragelogo}
            altText='logo'
            name='main-nav__logo'
          ></LeCarageLogo>
        </NavbarBrand>
        {/* <span className='main-nav__notofication-icon d-md-none'><BsBell /></span> */}
        <Nav className='ml-auto d-block' navbar>
          <NavItem style={{cursor:'pointer'}}>
            <NavLink data-testid='bsbell'>
              <BsBell />
            </NavLink>
          </NavItem>
          <UncontrolledDropdown setActiveFromChild>
            <DropdownToggle tag='a' className='nav-link' style={{cursor:'pointer'}}>
              My Profile
              <IoIosArrowDown className='nav-arrow-icon' />
            </DropdownToggle>
            <DropdownMenu>
              <DropdownItem tag='span' className='user-icon'>
                <AiOutlineUser />
                {parseJSON && parseJSON.email}
              </DropdownItem>
              <DropdownItem
                data-testid='dropdown-toggle'
                onClick={() => {
                  setShowLocationRole(!showLocationRole);
                }}
              >
                <div className='col-md-12 row'>
                  <div className='col-md-1'>
                    <SlEqualizer />
                  </div>{' '}
                  <div className='col-md-10 new'>{userProfile}</div>
                </div>
              </DropdownItem>
              <DropdownItem
                onClick={() => {
                  logOutOnClick();
                  dispatch(onLoggedOut());

                  //localStorage.removeItem('isAuthenticate');
                  // setAuth(false);
                  // navigate('/');
                }}
              >
                <FiLogOut /> LogOut
              </DropdownItem>
            </DropdownMenu>
          </UncontrolledDropdown>
        </Nav>
      </Navbar>
      {showLocationRole && (
        <ModalDropDown
          showModal={showLocationRole}
          onModalToggle={() => {
            setShowLocationRole(!showLocationRole);
          }}
          warehousesInfo={warehouses}
          profileInfo={profileInfo}
          handleUserProfile={handleUserProfile}
        />
      )}
    </div>
  );
};
