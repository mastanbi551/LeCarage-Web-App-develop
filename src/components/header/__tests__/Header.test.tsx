import React from 'react';
import { render, fireEvent, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Header } from '../Header';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { Store } from '../../../store/Store';

afterEach(cleanup);

const setAuth = jest.fn();

describe('Header component', () => {
  it('Header component should match snapshot', () => {
    const user = {
      id: 1,
      name: 'John Doe',
      email: 'johndoe@example.com',
    };
    const parseJSON = jest.fn().mockReturnValue(user);
    global.JSON.parse = parseJSON;
    const { container } = render(
      <Provider store={Store}>
        <BrowserRouter>
          <Header setAuth={setAuth} />
        </BrowserRouter>
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the component and display all elements', () => {
    const user = {
      id: 1,
      name: 'John Doe',
      email: 'johndoe@example.com',
    };
    const parseJSON = jest.fn().mockReturnValue(user);
    global.JSON.parse = parseJSON;
    render(
      <Provider store={Store}>
        <BrowserRouter>
          <Header setAuth={setAuth} />
        </BrowserRouter>
      </Provider>
    );
    const logoElement = screen.getByAltText('logo');
    expect(logoElement).toBeInTheDocument();
    const notificationIconElement = screen.getByTestId('bsbell');
    expect(notificationIconElement).toBeInTheDocument();
    const dropdownToggleElement = screen.getByText('My Profile');
    expect(dropdownToggleElement).toBeInTheDocument();
    const logoutElement = screen.getByText('LogOut');
    expect(logoutElement).toBeInTheDocument();
  });
});
