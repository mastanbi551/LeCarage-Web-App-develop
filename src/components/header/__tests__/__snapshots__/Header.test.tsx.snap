// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Header component should match snapshot 1`] = `
<div
  class="fixed"
>
  <nav
    class="main-nav navbar navbar-expand-md navbar-light bg-light"
  >
    <div
      class="container-fluid"
    >
      <a
        class="navbar-brand"
        href="/"
      >
        <img
          alt="logo"
          class="main-nav__logo"
          src="lecarageLogo.png"
        />
      </a>
      <ul
        class="ml-auto d-block navbar-nav"
      >
        <li
          class="nav-item"
          style="cursor: pointer;"
        >
          <a
            class="nav-link"
            data-testid="bsbell"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 16a2 2 0 0 0 2-2H6a2 2 0 0 0 2 2zM8 1.918l-.797.161A4.002 4.002 0 0 0 4 6c0 .628-.134 2.197-.459 3.742-.16.767-.376 1.566-.663 2.258h10.244c-.287-.692-.502-1.49-.663-2.258C12.134 8.197 12 6.628 12 6a4.002 4.002 0 0 0-3.203-3.92L8 1.917zM14.22 12c.223.447.481.801.78 1H1c.299-.199.557-.553.78-1C2.68 10.2 3 6.88 3 6c0-2.42 1.72-4.44 4.005-4.901a1 1 0 1 1 1.99 0A5.002 5.002 0 0 1 13 6c0 .88.32 4.2 1.22 6z"
              />
            </svg>
          </a>
        </li>
        <div
          class="dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="nav-link"
            style="cursor: pointer;"
          >
            My Profile
            <svg
              class="nav-arrow-icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 294.1L383 167c9.4-9.4 24.6-9.4 33.9 0s9.3 24.6 0 34L273 345c-9.1 9.1-23.7 9.3-33.1.7L95 201.1c-4.7-4.7-7-10.9-7-17s2.3-12.3 7-17c9.4-9.4 24.6-9.4 33.9 0l127.1 127z"
              />
            </svg>
          </a>
          <div
            aria-hidden="true"
            class="dropdown-menu"
            role="menu"
            tabindex="-1"
          >
            <span
              class="user-icon dropdown-item"
              role="menuitem"
              tabindex="0"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 1024 1024"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M858.5 763.6a374 374 0 0 0-80.6-119.5 375.63 375.63 0 0 0-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 0 0-80.6 119.5A371.7 371.7 0 0 0 136 901.8a8 8 0 0 0 8 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 0 0 8-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z"
                />
              </svg>
            </span>
            <button
              class="dropdown-item"
              data-testid="dropdown-toggle"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              <div
                class="col-md-12 row"
              >
                <div
                  class="col-md-1"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M160.048 387.872V32c0-17.664-14.336-32-32-32s-32 14.336-32 32v355.84C40.72 402.096-.352 452.288-.352 512c0 59.727 41.072 109.903 96.4 124.176V992c0 17.664 14.336 32 32 32s32-14.336 32-32V636.144c55.28-14.304 96.305-64.447 96.305-124.144 0-59.68-41.025-109.84-96.305-124.128zm-31.632 188.16c-.112 0-.24-.03-.368-.03-.144 0-.272.032-.415.048-35.153-.208-63.697-28.848-63.697-64.048 0-35.184 28.56-63.84 63.712-64.033.128 0 .272.033.4.033s.24-.033.368-.033c35.136.224 63.664 28.864 63.664 64.032 0 35.183-28.528 63.807-63.664 64.031zm415.648 3.842l-.001-547.872c0-17.664-14.336-32-32-32s-32 14.336-32 32v547.84c-55.328 14.273-96.4 64.433-96.4 124.16s41.072 109.903 96.4 124.176v163.824c0 17.664 14.336 32 32 32s32-14.336 32-32V828.146c55.264-14.304 96.288-64.447 96.288-124.144-.016-59.664-41.023-109.824-96.287-124.128zm-31.649 188.16c-.128 0-.24-.031-.368-.031-.144 0-.272.032-.415.048-35.153-.208-63.697-28.848-63.697-64.048 0-35.216 28.609-63.872 63.792-64.032.113 0 .225.032.337.032.096 0 .192-.032.288-.032 35.168.192 63.744 28.832 63.744 64.032 0 35.183-28.529 63.807-63.681 64.031zm415.648-572.159l-.001-163.871c0-17.664-14.336-32-32-32-17.68 0-32 14.336-32 32v163.84c-55.328 14.256-96.4 64.432-96.4 124.159s41.072 109.92 96.4 124.176v547.824c0 17.664 14.32 32 32 32 17.664 0 32-14.336 32-32V444.147c55.264-14.304 96.288-64.447 96.288-124.144-.016-59.664-41.023-109.824-96.287-124.128zm-31.649 188.161c-.128 0-.24-.031-.352-.031-.143 0-.287.032-.415.032-35.152-.192-63.712-28.832-63.712-64.032 0-35.184 28.56-63.84 63.696-64.032.143 0 .287.032.415.032s.24-.032.352-.032c35.136.24 63.664 28.864 63.664 64.032.033 35.183-28.496 63.807-63.648 64.031z"
                    />
                  </svg>
                </div>
                 
                <div
                  class="col-md-10 new"
                />
              </div>
            </button>
            <button
              class="dropdown-item"
              role="menuitem"
              tabindex="0"
              type="button"
            >
              <svg
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9 21H5a2 2 0 0 1-2-2V5a2 2 0 0 1 2-2h4"
                />
                <polyline
                  points="16 17 21 12 16 7"
                />
                <line
                  x1="21"
                  x2="9"
                  y1="12"
                  y2="12"
                />
              </svg>
               LogOut
            </button>
          </div>
        </div>
      </ul>
    </div>
  </nav>
</div>
`;
